modules/core/src/main/java/org/apache/ignite/internal/processors/service/GridServiceProcessor.java

```diff
@@ -220,217 +158,220 @@ public void onContinuousProcessorStarted(GridKernalContext ctx) throws IgniteChe
-        if (ctx.isDaemon())
+    @Override public void collectGridNodeData(DiscoveryDataBag dataBag) {
+        if (dataBag.commonDataCollectedFor(SERVICE_PROC.ordinal()))
             return;
 
-        GridSpinBusyLock busyLock = this.busyLock;
+        ServicesCommonDiscoveryData clusterData = new ServicesCommonDiscoveryData(
+            new ArrayList<>(registeredServices.values())
+        );
 
-        // Will not release it.
-        if (busyLock != null) {
-            busyLock.block();
+        dataBag.addGridCommonData(SERVICE_PROC.ordinal(), clusterData);
+    }
 
-            this.busyLock = null;
-        }
+    /** {@inheritDoc} */
+    @Override public void onGridDataReceived(DiscoveryDataBag.GridDiscoveryData data) {
+        if (data.commonData() == null)
+            return;
 
-        startLatch.countDown();
+        ServicesCommonDiscoveryData clusterData = (ServicesCommonDiscoveryData)data.commonData();
 
-        U.shutdownNow(GridServiceProcessor.class, depExe, log);
+        for (ServiceInfo desc : clusterData.registeredServices())
+            registeredServices.put(desc.serviceId(), desc);
+    }
 
-        if (!ctx.clientNode())
-            ctx.event().removeDiscoveryEventListener(topLsnr);
+    /** {@inheritDoc} */
+    @Override public void collectJoiningNodeData(DiscoveryDataBag dataBag) {
+        ArrayList<ServiceInfo> staticServicesInfo = staticallyConfiguredServices(true);
 
-        Collection<ServiceContextImpl> ctxs = new ArrayList<>();
+        dataBag.addJoiningNodeData(SERVICE_PROC.ordinal(), new ServicesJoinNodeDiscoveryData(staticServicesInfo));
+    }
 
-        synchronized (locSvcs) {
-            for (Collection<ServiceContextImpl> ctxs0 : locSvcs.values())
-                ctxs.addAll(ctxs0);
+    /** {@inheritDoc} */
+    @Override public void onJoiningNodeDataReceived(DiscoveryDataBag.JoiningNodeDiscoveryData data) {
+        if (data.joiningNodeData() == null)
+            return;
 
-            locSvcs.clear();
-        }
+        ServicesJoinNodeDiscoveryData joinData = (ServicesJoinNodeDiscoveryData)data.joiningNodeData();
 
-        for (ServiceContextImpl ctx : ctxs) {
-            ctx.setCancelled(true);
+        for (ServiceInfo desc : joinData.services()) {
+            assert desc.topologySnapshot().isEmpty();
 
-            Service svc = ctx.service();
+            ServiceInfo oldDesc = registeredServices.get(desc.serviceId());
 
-            if (svc != null)
-                try {
-                    svc.cancel(ctx);
-                }
-                catch (Throwable e) {
-                    log.error("Failed to cancel service (ignoring) [name=" + ctx.name() +
-                        ", execId=" + ctx.executionId() + ']', e);
+            if (oldDesc != null) { // In case of a collision of IgniteUuid.randomUuid() (almost impossible case)
+                U.warn(log, "Failed to register service configuration received from joining node : " +
```

Comment:
I think, debug level is enough here. This is a common situation, when every node has static service configuration on them. It would lead to a lot of warnings.

Reply:
The logging level has been chosen to be similar with IgniteCache.